### PR TITLE
Add hod-help-template command that will print diagnostic details.

### PR DESCRIPTION
**Pull request by ehiggs**
This tool prints out the resolution of the various templates to the command
line.

Run using: `./bin/hod-help-template`
The output looks like this on my machine:

```
$  bin/hod-help-template 
Hanythingondemand template parameters
Parameter name   :	Value                           	Documentation
dataaddress      :	157.193.44.242                  	Infiniband address if available
dataname         :	gast044b.ugent.be               	Infiniband hostname if available
hostaddress      :	157.193.44.242                  	IP address registered as the FQDN
hostname         :	squirtle                        	Fully Qualified Domain Name (FQDN)
localworkdir     :	WORKDIR/hod/ehiggs.squirtle.3538	Subdirectory of workdir with user, host, and pid in the name to make it distinct from other workdirs for use on shared file systems
masterdataaddress:	157.193.44.242                  	Address bound to the Infiniband adaptor on the master node if available
masterdataname   :	gast044b.ugent.be               	Hostname bound to the Infiniband adaptor on the master node if available
masterhostaddress:	157.193.44.242                  	Address bound to the Fully Qualified Domain Name (FQDN) of the master node.
masterhostname   :	squirtle                        	Hostname bound to the Fully Qualified Domain Name (FQDN) of the master node.
modules          :	MODULES                         	Modules listed in the hod.conf
pid              :	3538                            	PID for the current process
user             :	ehiggs                          	Current user
workdir          :	WORKDIR                         	Base directory for configuration and logging, e.g. /tmp, or somewhere on a shared file system.

```

### CODE REVIEW

File: hod/config/template.py [hunk 1]
```diff
@@ -88,7 +88,9 @@ def register_templates(template_registry, config_opts):
     local_data_network = node.sorted_network(node.get_networks())[0]
     templates = [
         _config_template_stub('masterhostname', 'Hostname bound to the Fully Qualified Domain Name (FQDN) of the master node.'),
+        _config_template_stub('masterhostaddress', 'Address bound to the Fully Qualified Domain Name (FQDN) of the master node.'),
         _config_template_stub('masterdataname', 'Hostname bound to the Infiniband adaptor on the master node if available'),
+        _config_template_stub('masterdataaddress', 'Address bound to the Infiniband adaptor on the master node if available'),
```
Comment (boegel) [inline hod/config/template.py]: this conflicts or reproduces what's in #26?

File: bin/hod-help-template [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    print 'Hanythingondemand template parameters'
+    longest_name = max(reg.fields.values(), key=lambda x: len(x.name)).name
+    max_name_len = len(longest_name)
+    value_names = map(resolver, ['$%s'% v.name for v in reg.fields.values()])
+    longest_value = max(value_names, key=len)
+    max_val_len = len(longest_value)
+    fmt_str = '%%-%ds:\t%%-%ds\t%%s' % (max_name_len, max_val_len)
+    print fmt_str % ('Parameter name', 'Value', 'Documentation')
+    for v in sorted(reg.fields.values(), key=lambda x: x.name):
+        print fmt_str % (v.name, resolver('$%s' % v.name), v.doc)
```
Comment (boegel) [inline bin/hod-help-template]: build up a `lines`, print in one go with `print `\n'.join(lines)`

File: bin/hod-help-template [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import socket
+import sys
+
+from hod.mpiservice import ConfigOptsParams, master_template_opts
+import hod.node.node as node
+
+def main(args):
+
+    config_opts = ConfigOptsParams(filename='hod.conf', workdir='WORKDIR',
+            modules=['MODULES'], master_template_kwargs=[])
```
Comment (boegel) [inline bin/hod-help-template]: why break this line?

File: bin/hod-help-template [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    reg = hct.TemplateRegistry()
+    hct.register_templates(reg, config_opts)
+    master_template_kwargs = master_template_opts(reg.fields.values())
+    for ct in master_template_kwargs:
+        reg.register(ct)
+    resolver = hct.TemplateResolver(**reg.to_kwargs())
+    print 'Hanythingondemand template parameters'
```
Comment (boegel) [inline bin/hod-help-template]: this is a long block, break it up a bit, add some comments?

File: hod/mpiservice.py [hunk 5]
```diff
@@ -122,17 +122,34 @@ def _slave_spread(comm):
...
+    fqdn = socket.getfqdn()
+    docs = dict()
+    if stub_config_opts is not None:
+        docs = dict([(opt.name, opt.doc) for opt in stub_config_opts])
+    return [
+            ConfigTemplate('masterhostname', fqdn, docs.get('masterhostname', '')),
+            ConfigTemplate('masterhostaddress', socket.gethostbyname(fqdn), docs.get('masterhostaddress', '')),
+            ConfigTemplate('masterdataname', master_dataname, docs.get('masterdataname', '')),
+            ConfigTemplate('masterdataaddress', master_dataaddress, docs.get('masterdataaddress', '')),
+            ]
```
Comment (boegel) [inline hod/mpiservice.py]: use 4-space indent?

File: test/unit/test_mpiservice.py [hunk 6]
```diff
@@ -77,6 +80,38 @@ def test_mpiservice_distribution(self):
...
+        self.assertEqual(opts[0].name, 'masterhostname')
+        self.assertEqual(opts[0].doc, '')
+        self.assertEqual(opts[1].name, 'masterhostaddress')
+        self.assertEqual(opts[1].doc, '')
+        self.assertEqual(opts[2].name, 'masterdataname')
+        self.assertEqual(opts[2].doc, '')
+        self.assertEqual(opts[3].name, 'masterdataaddress')
+        self.assertEqual(opts[3].doc, '')
+
+    def test_master_template_opts(self):
```
Comment (boegel) [inline test/unit/test_mpiservice.py]: you're overwriting the previous test here?

**Comment by boegel** [comment]
@ehiggs: Jenkins isn't configured yet to run the tests on HOD PRs?

File: bin/hod-help-template (hunk 2 above)
Comment (ehiggs) [inline bin/hod-help-template]: Why?

**Comment by ehiggs** [comment]
Jenkins isn't yet configured, no. I do have a `.travis.yml` file available.

File: bin/hod-help-template (hunk 3 above)
Comment (ehiggs) [inline bin/hod-help-template]: Because it's very long.

**Comment by boegel** [comment]
Jenkins: test this pleas

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by boegel** [comment]
Hmm, Jenkins trips over this:

```
    from mock import patch, MagicMock
.tox/py26/lib/python2.6/site-packages/mock/__init__.py:2: in <module>
    import mock.mock as _mock
E     File "/data/jenkins/workspace/hanythingondemand-pr-builder@3/TOXENV/py26/.tox/py26/lib/python2.6/site-packages/mock/mock.py", line 132
E       _builtins = {name for name in dir(builtins) if not name.startswith('_')}
E                           ^
E   SyntaxError: invalid syntax
```

@ehiggs: So, tox is not compatible with Python 2.6?

**Comment by ehiggs** [comment]
[Apparently the mock library for testing only support 2.7](https://pypi.python.org/pypi/mock/). The offending line is a dict comprehension which was indeed introduced in python 2.7 (or python 3.x and backported to 2.7).

**Comment by boegel** [comment]
Jenkins: test this please

**Pull request merged**